osu.Game/Screens/Spectate/SpectatorScreen.cs

```diff
@@ -62,26 +61,42 @@ protected override void LoadComplete()
         {
             base.LoadComplete();
 
-            spectatorClient.OnUserBeganPlaying += userBeganPlaying;
-            spectatorClient.OnUserFinishedPlaying += userFinishedPlaying;
-            spectatorClient.OnNewFrames += userSentFrames;
+            populateAllUsers().ContinueWith(_ => Schedule(() =>
+            {
+                spectatorClient.BindUserBeganPlaying(userBeganPlaying, true);
+                spectatorClient.OnUserFinishedPlaying += userFinishedPlaying;
+                spectatorClient.OnNewFrames += userSentFrames;
```

Comment:
I didn't realise this before but this subscription happens quite a bit later than it did previously. I'm thinking there could be a possibility of a `OnUserBeganPlaying`/`OnUserFinishedPlaying`/`OnNewFrames` event firing before `populateAllUsers()` has completed, which would lead to those events getting dropped.

I guess this was moved because it doesn't make sense to perform the `BindUserBeganPlaying(..., true)` call before users have been populated? Maybe best to hoist it out of the continuation and drop the `true`?

Reply:
Are you talking specifically about `OnNewFrames` here? If so, I don't think dropping some frames is considered a huge deal (and given the code just above it would ensure we don't receiver frames until a `UserBeganPlaying` has fired, which I *think* is what we want/expect).

Reply:
You're right that dropping frames is not a huge deal, but dropping a starting/finishing play event might be. Especially the finishing one - could lead to zombie entries if dropped, unless my thinking about things is entirely wrong.

Reply:
A start event can't be dropped, as it's `BindUserBeganPlaying(, true)`, right? I'm actually more concerned about event doubling because `BindUserBeganPlaying()` binds first and locks after...
As for the finishing event - I think you're right that there's a possibility it can be dropped - basically a thread race here. Perhaps binding the finishing event first?

Perhaps showcases an advantage with the way multiplayer does things as an alternative - everything's scheduled from there.

Will need to look into this again.

Reply:
Should have fixed these two issues - for the first, I'm now locking around the entire block. For the second I added another check in the `userBeganPlaying` callback to make sure the user is actually playing, by retrieving their current state.